[PyTorch/Distributed]Only sync buffers when broadcast_buffers is True

xing-liu opened the pull request:
Summary: Disable buffers sync in _sync_module_states(...) when broadcast_buffers is False. This change will memory usage when a model has huge buffers and does not need broadcast buffers.

Test Plan: .

Differential Revision: D45610709

mrshenli reviewed (approved):
fixes https://github.com/pytorch/pytorch/issues/99652

xing-liu committed:
[BC-Breaking][PyTorch/Distributed]: Only sync buffers when broadcast_buffers is True (#100729)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/100729

Disable buffers sync in _sync_module_states(...) when broadcast_buffers is False. This change will memory usage when a model has huge buffers and does not need broadcast buffers.

Test Plan: .

Differential Revision: D45610709

fbshipit-source-id: 20ce3f11e0a3d8375b4a70a8e4c78872442a8e51

xing-liu commented:
/easycla

mrshenli commented:
Test failure seems irrelevant

```
2023-05-06T23:19:21.0045862Z cuda eval  detectron2_fasterrcnn_r_101_c4      WARNING:common:fp64 golden ref were not generated for detectron2_fasterrcnn_r_101_c4. Setting accuracy check to cosine
2023-05-06T23:19:28.9728897Z Traceback (most recent call last):
2023-05-06T23:19:28.9731471Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/graph_module.py", line 271, in __call__
2023-05-06T23:19:28.9731875Z     return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
2023-05-06T23:19:28.9732334Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
2023-05-06T23:19:28.9732652Z     return self._call_impl(*args, **kwargs)
2023-05-06T23:19:28.9735960Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _call_impl
2023-05-06T23:19:28.9736383Z     return forward_call(*args, **kwargs)
2023-05-06T23:19:28.9737587Z   File "<eval_with_key>.7", line 8, in forward
2023-05-06T23:19:28.9738365Z     _assert_async = torch._assert_async(all_1, 'Box tensor contains infinite or NaN!');  all_1 = None
2023-05-06T23:19:28.9740545Z NotImplementedError: Could not run 'aten::_assert_async.msg' with arguments from the 'CUDA' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::_assert_async.msg' is only available for these backends: [CPU, Meta, BackendSelect, Python, FuncTorchDynamicLayerBackMode, Functionalize, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradHIP, AutogradXLA, AutogradMPS, AutogradIPU, AutogradXPU, AutogradHPU, AutogradVE, AutogradLazy, AutogradMeta, AutogradMTIA, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, AutogradNestedTensor, Tracer, AutocastCPU, AutocastCUDA, FuncTorchBatched, FuncTorchVmapMode, Batched, VmapMode, FuncTorchGradWrapper, PythonTLSSnapshot, FuncTorchDynamicLayerFrontMode, PythonDispatcher].
```
